File: python/naive.py
"""A naive approach to kNN, running in O(mndk) time."""

# Imports
import random
import time

# Parameters
N = 5000  # training set size
M = 200   # test set size
D = 20  # dimension of data
K = 5  # number of neighbours to use
SEED = 42  # random seed

# Helpers
def distance(X1, X2):
    """Euclidean distance between two vectors."""
    return sum(x1 * x2 for x1, x2 in zip(X1, X2))

# Generate dataset
random.seed(SEED)
X_train = [[random.gauss(0, 1) 
            for d in range(D)]
           for n in range(N)]
y_train = [random.gauss(0, 1) for n in range(N)]
X_test = [[random.gauss(0, 1) 
           for d in range(D)]
          for n in range(M)]
y_pred = []

# Record start time
start = time.time()
  
# kNN implementation
selected = [False for i in range(N)]
for X0 in X_test:
    neighbour_values = []
    for k in range(K):
        neighest = None  # (i, X, y, d)
        for i, (X, y) in enumerate(zip(X_train[1:], y_train[1:])):
            if selected[i]:
                continue
            d = distance(X0, X)
            if neighest is None or d < neighest[2]:
                neighest = i, X, y, d
        selected[i] = True
        neighbour_values.append(y)
    y_pred.append(sum(neighbour_values) / K)

# Report elapsed time
elapsed = time.time() - start
print(f"Completed in {elapsed:.02f} seconds")

File: java/Naive.java
// A naive approach to kNN, running in O(mndk) time.

import java.lang.Math;

public class Naive extends Base {

  public Naive(double[][] X_train, int[] y_train) {
    super(X_train, y_train);
  }

  public double[] predict(double[][] X_test, int K) {

    double[] pred = new double[0];

    for(int j=0; j<X_test.length; j++) {
      boolean[] selected = new boolean[this.N];
      double[] neighbour_values = new double[0];

      for(int k=0; k<K; k++) {
        int nearest_i = -1;
        double[] nearest_x = null;
        int nearest_y = -1;
        double nearest_d = -1;

        for(int i=0; i<this.X_train.length; i++) {
          if (selected[i]) {continue;}

          double d = this.dist(X_test[j], this.X_train[i]);

          if(((nearest_i == -1) && (nearest_y == -1) &&
              (nearest_x == null) && (nearest_d == -1)) || d < nearest_d) {

               nearest_i = i;
               nearest_x = this.X_train[i];
               nearest_y = this.y_train[i];
               nearest_d = d;
          }
        }
        selected[nearest_i] = true;

        // this is the irritating way to append values to an array in Java :(

        double[] temp = new double[(neighbour_values.length + 1)];
        for(int t=0; t<neighbour_values.length; t++) {
          temp[t] = neighbour_values[t];
        }
        temp[neighbour_values.length] = nearest_y;
        neighbour_values = temp;
      }
      double neighbour_sum = 0;

      for(int i=0; i<neighbour_values.length; i++) {
        neighbour_sum += neighbour_values[i];
      }

      neighbour_sum /= K;
      double[] temp2 = new double[(pred.length + 1)];
      for(int t=0; t<pred.length; t++) {
        temp2[t] = pred[t];
      }
      temp2[pred.length] = neighbour_sum;
      pred = temp2;
    }
    return pred;
  }

}
